contracts/vaults/Withdrawals.sol

```diff
@@ -0,0 +1,142 @@
+    * had also ended.
+    **/
+    function withdrawRequest() external nonReentrant {
+        HATVaultsRegistry.GeneralParameters memory generalParameters = registry.getGeneralParameters();
+        // require withdraw to be at least withdrawRequestEnablePeriod+withdrawRequestPendingPeriod since last withdrawRequest
+        // unless there's been a deposit or withdraw since, in which case withdrawRequest is allowed immediately
+        // solhint-disable-next-line not-rely-on-time
+        if (block.timestamp <
+            withdrawEnableStartTime[msg.sender] +
+                generalParameters.withdrawRequestEnablePeriod)
+            revert PendingWithdrawRequestExists();
+        // set the withdrawRequests time to be withdrawRequestPendingPeriod from now
+        // solhint-disable-next-line not-rely-on-time
+        withdrawEnableStartTime[msg.sender] = block.timestamp + generalParameters.withdrawRequestPendingPeriod;
+        emit WithdrawRequest(msg.sender, withdrawEnableStartTime[msg.sender]);
+    }
+
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares,
+        uint256 fee,
+        bool claimReward
+    ) internal nonReentrant {
+        // TODO: If a user gives allowance to another user, that other user can spam to some extent the allowing user's withdraw request
+        // Should consider disallowing withdraw from another user.
+        checkWithdrawAndResetWithdrawEnableStartTime(owner);
+        if (assets == 0) revert WithdrawMustBeGreaterThanZero();
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        rewardController.updateVaultBalance(owner, shares, false, claimReward);
+
+        _burn(owner, shares);
+
+        safeWithdrawVaultToken(assets, fee, receiver);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    /** @dev See {IERC4626-withdraw}. */
+    function withdraw(
+        uint256 assets,
+        address receiver,
+        address owner
+    ) public override virtual returns (uint256) {
+        if (assets > maxWithdraw(owner)) revert WithdrawMoreThanMax();
+
+        uint256 shares = previewWithdraw(assets);
+        uint256 fee = _convertToAssets(shares - _convertToShares(assets, MathUpgradeable.Rounding.Up), MathUpgradeable.Rounding.Up);
+        _withdraw(_msgSender(), receiver, owner, assets, shares, fee, true);
+
+        return shares;
+    }
+
+    /** @dev See {IERC4626-redeem}. */
+    function redeem(
+        uint256 shares,
+        address receiver,
+        address owner
+    ) public override virtual returns (uint256) {
+        if (shares > maxRedeem(owner)) revert RedeemMoreThanMax();
+
+        uint256 assets = previewRedeem(shares);
+        uint256 fee = _convertToAssets(shares, MathUpgradeable.Rounding.Down) - assets;
```

Comment:
Why not do:
```
uint256 assets = _convertToAssets(shares, MathUpgradeable.Rounding.Down);
uint256 fee = assets * withdrawalFee / HUNDRED_PERCENT;
 ```

Reply:
is a bit more explicit and also safes a bit of gas

Reply:
I'm not sure though I think the code is more consistent (with the withdraw function using preview withdraw), as it is now. You think we should change that anyway?